File: gaiag/runtime/cs/dzn/pump.cs
// Dezyne --- Dezyne command line tools
//
// Copyright © 2017 Rutger van Beusekom <rutger.van.beusekom@verum.com>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Diagnostics;
using System.Collections.Generic;

namespace dzn
{
  public class queue<T>:Queue<T>,IDisposable where T:IDisposable
  {
    public void Dispose()
    {
      foreach (T t in this) t.Dispose();
    }
  }

  public class pump : IDisposable
  {
    public static void port_block(Locator loc, Object p)
    {
      loc.get<pump>().block(p);
    }

    public static void port_release(Locator loc, Object p, Action out_binding)
    {
      if(out_binding!=null) out_binding();
      out_binding = null;
      loc.get<pump>().release(p);
    }

    public static coroutine find_self(list<coroutine> coroutines)
    {
      return coroutines.Find(c => c.port == null && !c.finished);
    }

    public static void finish(list<coroutine> coroutines)
    {
      coroutine self = find_self(coroutines);
      self.finished = true;
      Debug.WriteLine("[" + self.id + "] finish coroutine");
    }

    public Action worker;
    public Action next_event;
    public Action switch_context;
    public Action collateral_block_lambda;
    public Action exit;
    public list<coroutine> coroutines = new list<coroutine>();
    public list<coroutine> collateral_blocked = new list<coroutine>();
    public List<Object> skip_block = new List<Object>();
    public Queue<Action> queue = new Queue<Action>();
    public bool running;
    public Thread task;

    public pump()
    {
      this.collateral_block_lambda = () => {collateral_block();};
      this.running = true;
      this.task = new Thread(this.run);
      this.task.Start();
    }
    ~pump()
    {
      Dispose(false);
    }
    protected virtual void Dispose(bool gc)
    {
      stop();
      if(gc) {
        this.collateral_blocked.Dispose();
        this.coroutines.Dispose();
      }
    }
    public void Dispose()
    {
      Dispose(true);
      GC.SuppressFinalize(this);
    }
    public void stop()
    {
      context.lck(this, () => {
          this.running = false;
          Monitor.Pulse(this);
          Monitor.Exit(this);
          this.task.Join();
        });
    }
    public void wait()
    {
      context.lck(this, () => {
          while(this.queue.Count != 0) {
            Monitor.Wait(this);
          }
        });
    }
    public void run()
    {
      try {
        this.worker = () => {
          context.lck(this, () => {
              if(this.queue.Count == 0) {
                Monitor.Pulse(this);
              }
              while(this.queue.Count == 0 && running) {
                Monitor.Wait(this);
              }
              if(this.queue.Count != 0) {
                Action f = this.queue.Dequeue();
                Monitor.Exit(this);
                f();
              }
            });
        };

        coroutine zero = new coroutine();
        this.exit = ()=>{Debug.WriteLine("exit"); zero.release();};
        create_context();

        context.lck(this, () => {
            while(this.running || this.queue.Count!=0 || this.collateral_blocked.Count!=0)
            {
              Monitor.Exit(this);

              Debug.Assert(this.coroutines.Count!=0);

              this.coroutines.Last().call(zero);

              Monitor.Enter(this);

              this.coroutines.RemoveAll((c) => {
                  if(!c.finished) return false;
                  Debug.WriteLine("[" + c.id + "] removing");
                  c.Dispose();
                  return true;
                });
            }
            Debug.WriteLine("finish pump");
            Debug.Assert(this.queue.Count==0);
          });
      }
      catch(Exception e) {
        Console.Error.WriteLine("oops: " + e);
        System.Environment.Exit(1);
      }
    }
    public void create_context()
    {
      this.coroutines.Add (new coroutine (() =>
        {
          try
          {
            coroutine self = find_self(this.coroutines);
            Debug.WriteLine("[" + self.id + "] create context");
            while((this.running || this.queue.Count != 0) && !self.released)
            {
              worker();
              if(!self.released) collateral_release(self);
            }
            if(self.released) finish(this.coroutines);
            if(this.switch_context != null) {
              Action switch_context = this.switch_context;
              this.switch_context = null;
              switch_context();
            }
            if(!self.released) collateral_release(self);

            this.exit();
          }
          catch (forced_unwind) {
            Debug.WriteLine("ignoring forced_unwind");
          }
          catch (Exception e) {
            Console.Error.WriteLine("oops: " + e);
            System.Environment.Exit(1);
          }
        }));
    }
    public void collateral_block()
    {
      coroutine self = find_self(this.coroutines);
      Debug.WriteLine("[" + self.id + "] collateral_block");
      this.collateral_blocked.Add(self);
      this.coroutines.Remove(self);
      create_context();
      self.yield_to(this.coroutines.Last());
      Debug.WriteLine("[" + self.id + "] collateral_unblock");
    }
    public void collateral_release(coroutine self)
    {
      if(this.collateral_blocked.Count != 0) finish(coroutines);
      while(this.collateral_blocked.Count != 0) {
        this.coroutines.Add(this.collateral_blocked[0]);
        this.collateral_blocked.RemoveAt(0);
        self.yield_to(this.coroutines.Last());
      }
    }
    public void block(Object p)
    {
      int skip = this.skip_block.FindIndex(o => o == p);
      if(skip != -1) {
        this.skip_block.RemoveAt(skip);
        return;
      }

      coroutine self = find_self(this.coroutines);
      self.port = p;
      Debug.WriteLine("[" + self.id + "] block");
      create_context();
      self.yield_to(this.coroutines.Last());
      Debug.WriteLine("[" + self.id + "] re-entered context");
      Debug.Write("routines: ");
      foreach (coroutine c in this.coroutines){ Debug.Write(c.id + " ");}
      Debug.WriteLine("");

      this.coroutines.RemoveAll((c) => {
          if(!c.finished) return false;
          Debug.WriteLine("[" + c.id + "] removing");
          c.Dispose();
          return true;
        });
    }
    void release(Object p)
    {
      coroutine self = find_self(this.coroutines);
      coroutine blocked = this.coroutines.Find(c => c.port == p);
      if(blocked == null)
      {
        if (self!=null) Debug.WriteLine("[" + self.id + "] skip block");
        this.skip_block.Add(p);
        return;
      }

      Debug.WriteLine("[" + blocked.id + "] unblock");
      Debug.WriteLine("[" + self.id + "] released");
      self.released = true;

      this.switch_context = () => {
        blocked.port = null;
        Debug.WriteLine("[" + self.id + "] switch from");
        Debug.WriteLine("[" + blocked.id + "] to");
        self.yield_to(blocked);
      };
    }
    public void execute(Action e)
    {
      Debug.Assert(e != null);
      context.lck(this, () => {
          this.queue.Enqueue(e);
          Monitor.Pulse(this);
        });
    }
    public class promise: IDisposable
    {
      Barrier barrier;
      public promise()
      {
        barrier = new Barrier(2);
      }
      ~promise()
      {
        Dispose();
      }
      public void Dispose()
      {
        if(barrier != null) {
          barrier.Dispose();
          barrier = null;
        }
      }
      public void set()
      {
        barrier.SignalAndWait();
      }
      public void get()
      {
        barrier.SignalAndWait();
      }
    }
    public class promise<T>: promise
    {
      T value;
      public promise(): base()
      {}
      public void set(T value)
      {
        this.value = value;
        base.set();
      }
      public new T get()
      {
        base.get();
        return value;
      }
    }
    public void blocking(Action e)
    {
      using(promise p = new promise()) {
        this.execute(()=>{e(); p.set();});
        p.get();
      }
    }
    public T blocking<T>(Func<T> e)
    {
      using(promise<T> p = new promise<T>()) {
        this.execute(()=>{p.set(e());});
        return p.get();
      }
    }
  }
}

File: gaiag/runtime/c++03/pump.cc
// Dezyne --- Dezyne command line tools
//
// Copyright © 2015, 2016, 2017, 2018 Rutger van Beusekom <rutger.van.beusekom@verum.com>
// Copyright © 2016 Henk Katerberg <henk.katerberg@yahoo.com>
// Copyright © 2015, 2016 Jan Nieuwenhuizen <janneke@gnu.org>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

#include <dzn/locator.hh>
#include <dzn/meta.hh>
#include <dzn/pump.hh>

#include <algorithm>
#include <cassert>
#include <new>

#ifdef DEBUG_RUNTIME
#include <iostream>
#endif

namespace dzn
{
  size_t coroutine::g_id = 0;

  void port_block(const locator& l, void* p)
  {
    l.get<dzn::pump>().block(p);
  }
  void port_release(const locator& l, void* p, boost::function<void()>& out_binding)
  {
    if(out_binding) out_binding();
    out_binding = 0;
    l.get<dzn::pump>().release(p);
  }
  void collateral_block(const locator& l)
  {
    l.get<dzn::pump>().collateral_block_lambda();
  }
  inline void compose_sequence(const boost::function<void()>& f, const boost::function<void()>& g) {f(); g();}
  void call_async_helper(const locator& l, const meta& m, size_t id, const boost::function<void()>& f, const boost::function<void()>& g)
  {
    l.get<dzn::pump>().handle(id, 0, boost::bind(compose_sequence,f,g), m.rank);
  }

  inline bool unblocked_p(const coroutine& c)
  {
    return c.port == 0 && !c.finished;
  }

  static std::list<coroutine>::iterator find_self(std::list<coroutine>& coroutines)
  {
    assert(1 == std::count_if(coroutines.begin(), coroutines.end(), unblocked_p));
    std::list<coroutine>::iterator self = std::find_if(coroutines.begin(), coroutines.end(), unblocked_p);
    return self;
  }

  inline bool blocked_p(void* port, const coroutine& c)
  {
    return port == c.port;
  }

  static std::list<coroutine>::iterator find_blocked(std::list<coroutine>& coroutines, void* port)
  {
    std::list<coroutine>::iterator self = std::find_if(coroutines.begin(), coroutines.end(), boost::bind<bool>(blocked_p, port, _1));
    return self;
  }

  static void finish(std::list<coroutine>& coroutines)
  {
    std::list<coroutine>::iterator self = find_self(coroutines);
    self->finished = true;
    debug << "[" << self->id << "] finish coroutine" << std::endl;
  }

  pump::pump()
  : collateral_block_lambda(boost::bind(&pump::collateral_block, this))
  , switch_context()
  , running(true)
  , task(boost::async(boost::launch::async, boost::function<void()>(boost::ref(*this))))
  {}
  pump::~pump()
  {
    stop();
  }
  void pump::stop()
  {
    boost::unique_lock<boost::mutex> lock(mutex);
    if(running)
    {
      running = false;
      condition.notify_one();
      if (lock) lock.unlock();
      task.wait();
    }
  }
  void pump::wait()
  {
    boost::unique_lock<boost::mutex> lock(mutex);
    while(queue.empty()) {
      idle.wait(lock);
    }
  }
  void pump::worker_helper()
  {
    boost::unique_lock<boost::mutex> lock(mutex);
    if(queue.empty())
    {
      idle.notify_one();
    }
    if(timers.empty())
    {
      while(queue.empty() && running)
        condition.wait(lock);
    }
    else
    {
      while(queue.empty() && running)
        condition.wait_until(lock, timers.begin()->first.t);
    }

    if(queue.size())
    {
      boost::function<void()> f;
      std::swap(queue.front(), f);
      queue.pop();
      lock.unlock();
      f();
      lock.lock();
    }

    while(timers.size() && timers.begin()->first.expired())
    {
      boost::function<void()> f(timers.begin()->second);
      timers.erase(timers.begin());
      lock.unlock();
      f();
      lock.lock();
    }
  }
  void pump::operator()()
  {
    try
    {
      thread_id = boost::this_thread::get_id();

      worker = boost::bind(&pump::worker_helper, this);

      coroutine zero;
      create_context();

      exit = boost::bind(&coroutine::release, &zero);

      boost::unique_lock<boost::mutex> lock(mutex);
      while(running || queue.size() || collateral_blocked.size())
      {
        assert(coroutines.size());

        lock.unlock();

        coroutines.back().call(zero);

        lock.lock();

        remove_finished_coroutines();
      }
      debug << "finish pump" << std::endl;
      assert(queue.empty());
    }
    catch(const std::exception& e)
    {
      debug << __FILE__ << ":" << __LINE__ << ": oops: " << e.what() << std::endl;
      std::terminate();
    }
  }
  void pump::create_context_helper()
  {
    try
    {
      std::list<coroutine>::iterator self = find_self(coroutines);
      debug << "[" << self->id << "] create context" << std::endl;
      while(!self->released && (running ||
                                queue.size() ||
                                coroutines.size() > 1))
      {
        worker();
        if(!self->released) collateral_release(self);
      }
      if(self->released) finish(coroutines);

      if(switch_context) {
        boost::function<void()> tmp;
        std::swap(tmp, switch_context);
        tmp();
      }

      if(!self->released) collateral_release(self);

      exit();
    }
    catch(const forced_unwind&) {throw;}
    catch(const std::exception& e)
    {
      debug << __FILE__ << ":" << __LINE__ << ": oops: " << e.what() << std::endl;
      std::terminate();
    }
  }
  void pump::create_context()
  {
    coroutines.push_back(coroutine());
    coroutines.back().~coroutine();
    new (&coroutines.back()) dzn::coroutine(boost::function<void()>(boost::bind(&pump::create_context_helper, this)));
  }
  void pump::collateral_block()
  {
    std::list<coroutine>::iterator self = find_self(coroutines);
    debug << "[" << self->id << "] collateral_block" << std::endl;

    collateral_blocked.splice(collateral_blocked.end(), coroutines, self);
    create_context();
    self->yield_to(coroutines.back());

    debug << "[" << self->id << "] collateral_unblock" << std::endl;
  }
  void pump::collateral_release(std::list<coroutine>::iterator self)
  {
    if(collateral_blocked.size()) finish(coroutines);
    while(collateral_blocked.size())
    {
      coroutines.splice(coroutines.end(), collateral_blocked, collateral_blocked.begin());
      self->yield_to(coroutines.back());
    }
  }
  void pump::block(void* p)
  {
    std::set<void*>::iterator it = skip_block.find(p);
    if(it != skip_block.end())
    {
      skip_block.erase(it);
      return;
    }

    std::list<coroutine>::iterator self = find_self(coroutines);

    self->port = p;

    debug << "[" << self->id << "] block" << std::endl;

    create_context();

    self->yield_to(coroutines.back());
    debug << "[" << self->id << "] entered context" << std::endl;
    if (debug.rdbuf())
    {
      debug << "routines: ";
      for (std::list<coroutine>::iterator it = coroutines.begin(); it != coroutines.end(); ++it) {
        debug << it->id << " ";
      }
      debug << std::endl;
    }
    remove_finished_coroutines();
  }
  inline void switch_context_helper(std::list<coroutine>::iterator blocked, std::list<coroutine>::iterator self)
  {
    blocked->port = 0;

    debug << "[" << self->id << "] switch from" << std::endl;
    debug << "[" << blocked->id << "] to" << std::endl;

    self->yield_to(*blocked);
  }
  void pump::release(void* p)
  {
    std::list<coroutine>::iterator self = find_self(coroutines);

    std::list<coroutine>::iterator blocked = find_blocked(coroutines, p);
    if(blocked == coroutines.end())
    {
    debug << "[" << self->id << "] skip block" << std::endl;
      skip_block.insert(p);
      return;
    }

    debug << "[" << blocked->id << "] unblock" << std::endl;
    debug << "[" << self->id << "] released" << std::endl;
    self->released = true;

    switch_context = boost::bind(&switch_context_helper,blocked,self);
  }
  void pump::operator()(const boost::function<void()>& e)
  {
    assert(e);
    boost::lock_guard<boost::mutex> lock(mutex);
    queue.push(e);
    condition.notify_one();
  }
  inline bool timer_p(size_t id, const std::pair<pump::deadline, boost::function<void()> >& p)
  {
    return p.first.id == id;
  }
  void pump::handle(size_t id, size_t ms, const boost::function<void()>& e, size_t rank)
  {
    assert(e);
    assert(std::find_if(timers.begin(), timers.end(), boost::bind<bool>(&timer_p, id, _1)) == timers.end());
    timers.insert(std::make_pair(deadline(id, ms, rank), e));
  }
  void pump::remove(size_t id)
  {
    std::map<deadline, boost::function<void()> >::iterator it = std::find_if(timers.begin(), timers.end(), boost::bind<bool>(&timer_p, id, _1));
    if(it != timers.end()) timers.erase(it);
  }

  inline bool finished_p(const coroutine& c){if(c.finished) debug << "[" << c.id << "] removing" << std::endl; return c.finished;}

  void pump::remove_finished_coroutines()
  {
    coroutines.remove_if(finished_p);
  }
}
